synadm/api.py

```diff
@@ -289,6 +290,31 @@ def server_name(self):
             return None
         return resp['server_name']
 
+    def generate_mxid(self, user_id):
+        """ Checks whether the given user ID is an MXID already or else
+        generates it from the passed string and the homeserver name fetched
+        via the server_name method.
+
+        Args:
+            user_id (string): User ID given by user as command argument.
+
+        Returns:
+            string: the fully qualified Matrix User ID (MXID) or None if the
+                user_id parameter is None.
+        """
+        if user_id is None:
+            return None
+        elif any(x in user_id for x in ["@", ":"]):
+            if re.match(r"\@.+\:.+", user_id):
```

Comment:
I don't think the @ and the : need to be escaped. Would improve readability a little IMHO.

I like that it's even recognized when @localpart or localpart: is the input! Nice!

Also we could narrow down the regex match even more. The problem is I just can't find documentation about which characters are actually allowed in the localpart. Is dash and underscore allowed? I just asked that on #synapse:matrix.org.

Anyway, what shoud definitely work but probabIy would be a little too much allowed still is using the "non whitespace sequence `\S`": `@\S+:\S`

https://docs.python.org/3.7/library/re.html

Reply:
I agree that the readability is currently not ideal. 

Regarding the rexeg match: The specification of the MXID can be found here: https://spec.matrix.org/v1.2/appendices/#user-identifiers. As stated in the specification, we have to include the following characters: `a-z`, `0-9`, `.`, `_`, `=`, `-`, and `/` as valid input. Therefore I think that the blanket use of `\S` would not be ideal since it also accepts characters like `?` or `+` which are not allowed within the localpart. 
For input validation it would be best to explicitly specify the allowed character, e.g. by using the regex `[\w\s\_\.\_\=\-\/]+` for the localpart. I know that this regex is probably not the best regarding readability but it catches only the allowed characters. Let me know what you think about this approach.

Reply:
Go for it! Specific is good! 
I think the \s (whitespace) slipped in and shouldnt be there, right?

Reply:
I have implemented this approach and improved the strcuture of the MXID generation in 406d5792abc9b277c4b3c3741e3db9de01f4fd04. I have already done basic testing, but more in-depth testing is required. Please let me know if you find any issues.